fhir-persistence-jdbc/src/main/java/com/ibm/fhir/persistence/jdbc/util/TimestampPrefixedUUID.java

```diff
@@ -0,0 +1,42 @@
+/*
+ * (C) Copyright IBM Corp. 2020
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.ibm.fhir.persistence.jdbc.util;
+
+import java.util.UUID;
+
+/**
+ * Provides identity strings using random UUID for uniqueness but
+ * prefixed with an encoded time string to improve database locality
+ * when used in b-tree indexes.
+ */
+public class TimestampPrefixedUUID implements LogicalIdentityProvider {
+
+    @Override
+    public String createNewIdentityValue() {
+        // It's OK to use milli-time here. It doesn't matter too much if the time changes
+        // because we're not using the timestamp to determine uniqueness in any way. The
+        // timestamp prefix is purely to help push index writes to the right hand side
+        // of the btree, minimizing the number of physical reads likely required 
+        // during ingestion when an index is too large to be fully cached.
+        long millis = System.currentTimeMillis();
+        
+        // String encoding. Needs to collate correctly, so don't use any
+        // byte-based encoding which would be sensitive to endian issues. For simplicity,
+        // hex is sufficient, although a custom encoding using the full character set
+        // supported by FHIR identifiers would be a little more compact (== smaller indexes).
+        // Do not use Base64.
+        String prefix = Long.toHexString(millis);
+        
+        UUID uuid = UUID.randomUUID();
+        
+        StringBuilder result = new StringBuilder();
+        result.append(prefix);
+        result.append("-"); // redundant, but more visually appealing.
```

Comment:
Does this impact any regular expressions? or potentially any fhir path? 

Does it impact FHIR Uuid type? 
`private static final Pattern PATTERN = Pattern.compile("urn:uuid:[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}");`.

Reply:
Resource.id is a system string with a hinted type of https://www.hl7.org/fhir/datatypes.html#id (and not UUID), so as long as we only use it for that, then I think we're OK on this one.

Reply:
It shouldn't, because identifiers are constrained by a different pattern according to the spec.

Reply:
Works for me